lte/gateway/c/session_manager/LocalEnforcer.cpp

```diff
@@ -304,126 +305,180 @@ void LocalEnforcer::execute_actions(
+                action_p->get_rule_definitions(), _1, _2));
+        break;
+      case REDIRECT:
+        // This is GY based REDIRECT, GX redirect will come in as a regular rule
+        install_redirect_flow(action_p, session_update);
+        break;
+      case RESTRICT_ACCESS:
+        MLOG(MWARNING) << "RESTRICT_ACCESS mode is unsupported"
+                       << ", will just terminate the service.";
+      case TERMINATE_SERVICE: {
+        auto it = session_map.find(imsi);
+        if (it == session_map.end()) {
+          MLOG(MERROR) << "Cannot act on TERMINATE action since session "
+                       << session_id << " does not exist";
+          break;
+        }
+        for (const auto& session : it->second) {
+          if (session->get_session_id() == session_id) {
+            auto& uc = session_update[imsi][session_id];
+            start_session_termination(imsi, session, true, uc);
+            break;
+          }
+        }
+        // Session not found
+        MLOG(MERROR) << "Cannot act on TERMINATE action since session "
+                     << session_id << " does not exist";
+        break;
+      }
+      case CONTINUE_SERVICE:
+        break;
     }
   }
 }
 
-// Terminates sessions that correspond to the given IMSI.
-// (For session termination triggered by sessiond)
-void LocalEnforcer::terminate_service(
-    SessionMap& session_map, const std::string& imsi,
-    const std::vector<std::string>& rule_ids,
-    const std::vector<PolicyRule>& dynamic_rules,
-    SessionUpdate& session_update) {
-  pipelined_client_->deactivate_flows_for_rules(
-      imsi, rule_ids, dynamic_rules, RequestOriginType::GX);
-  MLOG(MINFO) << "Initiating session termination for " << imsi;
-
-  auto it = session_map.find(imsi);
-  if (it == session_map.end()) {
-    MLOG(MWARNING) << "Could not find session with IMSI " << imsi
-                   << " to terminate";
+// Terminates sessions that correspond to the given IMSI and session.
+void LocalEnforcer::start_session_termination(
+    const std::string& imsi, const std::unique_ptr<SessionState>& session,
+    bool notify_access, SessionStateUpdateCriteria& update_criteria) {
+  auto session_id = session->get_session_id();
+  if (session->is_terminating()) {
+    // If the session is terminating already, do nothing.
+    MLOG(MDEBUG) << "Session " << session_id << " is already terminating, "
+                 << "ignoring termination request";
     return;
   }
+  MLOG(MINFO) << "Initiating session termination for " << session_id;
 
-  for (const auto& session : it->second) {
-    if (session->is_terminating()) {
-      // If the session is terminating already, do nothing.
-      continue;
+  remove_all_rules_for_termination(imsi, session, update_criteria);
+
+  session->set_fsm_state(SESSION_TERMINATING_FLOW_ACTIVE, update_criteria);
+  auto config = session->get_config();
+  if (notify_access) {
+    notify_termination_to_access(imsi, session_id, config);
```

Comment:
Good call, will rename to `notify_termination_to_access_service`. And yes, this notifies either AAA or MME with termination, so that they can detach the user.